File: jflove-server/src/utils/crypto.py
"""
加密工具模块

提供 X25519 密钥对生成、ECDH+HKDF 会话密钥派生、
ChaCha20-Poly1305 对称加解密能力。

包括两套对称加密 API：
  - 整包加密（encrypt / decrypt）：用于普通 JSON 请求/响应
  - 流式分帧加密（encrypt_stream_chunk）：用于文件下载/预览的响应体，
    每片独立 nonce，配合客户端 decrypt_stream_chunk 实现端到端加密文件流
"""

import os
import struct
import base64
from cryptography.hazmat.primitives.asymmetric.x25519 import X25519PrivateKey
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305
from src.config.settings import SESSION_KEY_SALT

# 流式分帧大小（明文字节数）；密文帧 = 4B 长度 + 12B nonce + plaintext + 16B tag
STREAM_PLAINTEXT_CHUNK_SIZE = 64 * 1024


def generate_x25519_keypair() -> tuple[X25519PrivateKey, str]:
    """
    生成 X25519 临时密钥对。

    :returns: (私钥对象, 公钥 Base64 字符串)
    """
    private_key = X25519PrivateKey.generate()
    public_key_bytes = private_key.public_key().public_bytes(
        serialization.Encoding.Raw,
        serialization.PublicFormat.Raw,
    )
    return private_key, base64.b64encode(public_key_bytes).decode()


def derive_session_key(private_key: X25519PrivateKey, peer_public_key_b64: str) -> bytes:
    """
    通过 ECDH 计算共享密钥，再经 HKDF-SHA256 派生 32 字节会话密钥。

    :param private_key: 本端 X25519 私钥对象
    :param peer_public_key_b64: 对端公钥 Base64 字符串
    :returns: 32 字节会话密钥
    """
    peer_pub_bytes = base64.b64decode(peer_public_key_b64)
    from cryptography.hazmat.primitives.asymmetric.x25519 import X25519PublicKey
    peer_pub = X25519PublicKey.from_public_bytes(peer_pub_bytes)
    shared_secret = private_key.exchange(peer_pub)

    hkdf = HKDF(
        algorithm=hashes.SHA256(),
        length=32,
        salt=SESSION_KEY_SALT,
        info=b"",
    )
    return hkdf.derive(shared_secret)


def encrypt(session_key: bytes, plaintext: bytes) -> dict:
    """
    使用 ChaCha20-Poly1305 加密明文。

    每次加密随机生成 12 字节 nonce，保证密文唯一性。

    :param session_key: 32 字节会话密钥
    :param plaintext: 待加密的原始字节
    :returns: {"nonce": "<Base64>", "ciphertext": "<Base64>"}
    """
    nonce = os.urandom(12)
    chacha = ChaCha20Poly1305(session_key)
    ciphertext = chacha.encrypt(nonce, plaintext, None)
    return {
        "nonce": base64.b64encode(nonce).decode(),
        "ciphertext": base64.b64encode(ciphertext).decode(),
    }


def decrypt(session_key: bytes, nonce_b64: str, ciphertext_b64: str) -> bytes:
    """
    使用 ChaCha20-Poly1305 解密密文，同时验证认证标签。

    :param session_key: 32 字节会话密钥
    :param nonce_b64: nonce Base64 字符串
    :param ciphertext_b64: 密文 Base64 字符串（含认证标签）
    :returns: 解密后的原始字节
    :raises: cryptography.exceptions.InvalidTag 若认证失败
    """
    nonce = base64.b64decode(nonce_b64)
    ciphertext = base64.b64decode(ciphertext_b64)
    chacha = ChaCha20Poly1305(session_key)
    return chacha.decrypt(nonce, ciphertext, None)


def encrypt_stream_chunk(session_key: bytes, plaintext: bytes) -> bytes:
    """
    流式分帧加密：把一片明文加密为一帧二进制，可直接写入 HTTP 流式响应。

    帧格式（大端字节序）：
        [4B 帧体长度 L] [12B nonce] [N B 密文 + 16B 认证标签]
    其中 L = 12 + N + 16 = 28 + 明文字节数

    每帧使用独立随机 nonce，ChaCha20-Poly1305 自带认证标签可防篡改与截断。

    :param session_key: 32 字节会话密钥（与请求/响应整包加密共用）
    :param plaintext: 明文字节（建议每帧 STREAM_PLAINTEXT_CHUNK_SIZE）
    :returns: 完整帧字节流，可直接 yield 给 StreamingResponse
    """
    nonce = os.urandom(12)
    chacha = ChaCha20Poly1305(session_key)
    ciphertext = chacha.encrypt(nonce, plaintext, None)
    body = nonce + ciphertext  # 12 + len(plaintext) + 16
    return struct.pack(">I", len(body)) + body

File: jflove-web/src/utils/crypto.ts
/**
 * 加密工具模块
 *
 * 对标桌面端 src/utils/crypto.py 和移动端 lib/utils/crypto.dart。
 * 使用 Web Crypto API (X25519 ECDH + HKDF-SHA256) + @noble/ciphers (ChaCha20-Poly1305)。
 *
 * 加密协议三端统一：
 *   - 密钥交换：X25519 ECDH
 *   - 密钥派生：HKDF-SHA256（盐 b"jflove-v1"，32B）
 *   - 数据加密：ChaCha20-Poly1305（12B 随机 nonce）
 *   - 加密信封：{"nonce": "<Base64>", "ciphertext": "<Base64>"}
 *   - 流式帧：  [4B 大端长度][12B nonce][密文+16B Poly1305 tag]
 */

import { chacha20poly1305 } from '@noble/ciphers/chacha';
import { SESSION_KEY_SALT, SESSION_KEY_LENGTH, NONCE_LENGTH } from '../config/constants';

// ── X25519 密钥生成 ───────────────────────────

/** 生成 X25519 临时密钥对 */
export async function generateKeyPair(): Promise<{
  publicKeyB64: string;
  privateKey: CryptoKey;
}> {
  const keyPair = await crypto.subtle.generateKey(
    { name: 'X25519' },
    true,
    ['deriveBits'],
  ) as CryptoKeyPair;
  const publicKeyRaw = await crypto.subtle.exportKey('raw', keyPair.publicKey);
  const publicKeyB64 = arrayBufferToBase64(publicKeyRaw);
  return { publicKeyB64, privateKey: keyPair.privateKey };
}

// ── ECDH + HKDF 派生 session_key ──────────────

/** ECDH + HKDF-SHA256 派生 32 字节会话密钥 */
export async function deriveSessionKey(
  privateKey: CryptoKey,
  peerPublicKeyB64: string,
): Promise<Uint8Array> {
  const peerPublicKeyRaw = base64ToUint8Array(peerPublicKeyB64);

  const peerPublicKey = await crypto.subtle.importKey(
    'raw',
    peerPublicKeyRaw.buffer as ArrayBuffer,
    { name: 'X25519' },
    false,
    [],
  );

  // ECDH 共享密钥
  const sharedSecret = await crypto.subtle.deriveBits(
    {
      name: 'X25519',
      public: peerPublicKey,
    },
    privateKey,
    256, // X25519 输出 256 bits
  );

  // HKDF-SHA256 派生 32 字节 session_key
  const hkdfKey = await crypto.subtle.importKey(
    'raw',
    new Uint8Array(sharedSecret),
    { name: 'HKDF' },
    false,
    ['deriveBits'],
  );

  const sessionKey = await crypto.subtle.deriveBits(
    {
      name: 'HKDF',
      hash: 'SHA-256',
      salt: SESSION_KEY_SALT,
      info: new Uint8Array(0),
    },
    hkdfKey,
    SESSION_KEY_LENGTH * 8,
  );

  return new Uint8Array(sessionKey);
}

// ── ChaCha20-Poly1305 加密/解密 ───────────────

/** ChaCha20-Poly1305 加密 */
export function encryptEnvelope(
  sessionKey: Uint8Array,
  plaintext: Uint8Array,
): { nonce: string; ciphertext: string } {
  const nonce = crypto.getRandomValues(new Uint8Array(NONCE_LENGTH));
  const chacha = chacha20poly1305(sessionKey, nonce);
  const ciphertext = chacha.encrypt(plaintext);
  return {
    nonce: arrayBufferToBase64(nonce),
    ciphertext: arrayBufferToBase64(ciphertext),
  };
}

/** ChaCha20-Poly1305 解密 */
export function decryptEnvelope(
  sessionKey: Uint8Array,
  nonceB64: string,
  ciphertextB64: string,
): Uint8Array {
  const nonce = base64ToUint8Array(nonceB64);
  const ciphertext = base64ToUint8Array(ciphertextB64);
  const chacha = chacha20poly1305(sessionKey, nonce);
  return chacha.decrypt(ciphertext);
}

// ── 流式帧加密/解密 ─────────────────────────

/**
 * 加密流式帧。
 * 帧格式：[4B 大端长度][12B nonce][密文+16B Poly1305 tag]
 * 长度 = 12 + plaintext.length + 16（nonce + ciphertext + tag）
 */
export function encryptStreamChunk(
  sessionKey: Uint8Array,
  plaintext: Uint8Array,
): Uint8Array {
  const nonce = crypto.getRandomValues(new Uint8Array(NONCE_LENGTH));
  const chacha = chacha20poly1305(sessionKey, nonce);
  const ciphertext = chacha.encrypt(plaintext);

  // 帧总长度 = 4B 长度 + 12B nonce + ciphertext
  const frameLength = NONCE_LENGTH + ciphertext.length;
  const result = new Uint8Array(4 + frameLength);

  // 写入 4 字节大端长度
  const view = new DataView(result.buffer);
  view.setUint32(0, frameLength, false); // big-endian

  result.set(nonce, 4);
  result.set(ciphertext, 4 + NONCE_LENGTH);

  return result;
}

/** 解密流式帧 */
export function decryptStreamChunk(
  sessionKey: Uint8Array,
  frameBody: Uint8Array,
): Uint8Array {
  const nonce = frameBody.slice(0, NONCE_LENGTH);
  const ciphertext = frameBody.slice(NONCE_LENGTH);
  const chacha = chacha20poly1305(sessionKey, nonce);
  return chacha.decrypt(ciphertext);
}

// ── Base64 / Uint8Array 转换 ────────────────

export function arrayBufferToBase64(buffer: ArrayBuffer | Uint8Array): string {
  const bytes = buffer instanceof Uint8Array ? buffer : new Uint8Array(buffer);
  let binary = '';
  for (let i = 0; i < bytes.length; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  return btoa(binary);
}

export function base64ToUint8Array(b64: string): Uint8Array {
  const binary = atob(b64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
}

// ── BufferSource → Uint8Array ───────────────

export function cryptoKeyToUint8Array(key: ArrayBuffer): Uint8Array {
  return new Uint8Array(key);
}
